internal/zipper/zipper_test.go

```diff
@@ -0,0 +1,130 @@
+package zipper
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/composite-action-framework-go/pkg/fs"
+)
+
+type files map[string]string
+
+func TestZipper_ZipDir_ok(t *testing.T) {
+
+	cases := []files{
+		{
+			"text.txt": "hello!",
+		},
+		{
+			"subdir/text.txt": "hello!",
+		},
+		{
+			"subdir/text.txt":         "hello!",
+			"subdir/blah/blah/hi.txt": "hello!",
+		},
+	}
+
+	for _, c := range cases {
+		inputFiles := c
+		t.Run("", func(t *testing.T) {
+			buf := &bytes.Buffer{}
+			z := New(buf)
+
+			dir := createTestDir(t, inputFiles)
+			if err := z.ZipDir(dir); err != nil {
+				t.Fatal(err)
+			}
+
+			byteReader := bytes.NewReader(buf.Bytes())
+			reader, err := zip.NewReader(byteReader, int64(buf.Len()))
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			wantFiles := make(files, len(inputFiles))
+			for name := range inputFiles {
+				// We expect to see a flattened hierarchy here,
+				// so throw away the dir component.
+				wantFiles[filepath.Base(name)] = ""
+			}
+
+			for _, f := range reader.File {
+				_, ok := wantFiles[f.Name]
+				if !ok {
+					t.Errorf("unexpected file %q", f.Name)
+					continue
+				}
+				delete(wantFiles, f.Name)
+				t.Logf("Zip contains %q", f.Name)
+			}
+
+			for name := range wantFiles {
+				t.Errorf("file %q missing from zip", name)
+			}
+
+		})
+	}
+}
+
+func TestZipper_ZipDir_err_duplicate(t *testing.T) {
+
+	cases := []files{
+		{
+			"text.txt":        "hello!",
+			"subdir/text.txt": "hello!",
+		},
+		{
+			"subdir/text.txt":      "hello!",
+			"subdir/blah/text.txt": "hello!",
+		},
+	}
+
+	for _, c := range cases {
+		inputFiles := c
+		t.Run("", func(t *testing.T) {
+			buf := &bytes.Buffer{}
+			z := New(buf)
+
+			dir := createTestDir(t, inputFiles)
+			err := z.ZipDir(dir)
+			if err == nil {
+				t.Fatalf("got nil error; want duplicate entry error")
+			}
+			if !strings.Contains(err.Error(), "duplicate") {
+				t.Fatalf("got error %q; want duplicate entry error", err)
+			}
+		})
+	}
+
+}
+
+func createTestDir(t *testing.T, f files) string {
```

Comment:
This function takes a `files` and writes it to disk in a temp directory so we can run the tests against the real filesystem. It would be nice to use a mock filesystem for this instead if we revisit this test.